### DISCUSSION TITLE
Find all cycles in k-partite directed graph

### DISCUSSION THREAD

**Discussion by jeiros** [networkx/networkx#5332] [Q&A] [answered]
Apologies if this usability question is not meant for this forum - please let me know if this is better suited for the mailing list.

I must find *all* the cycles in a k-partite (I think that's the right term) directed graph. Consider the following graph:

```python

import networkx as nx
G = nx.DiGraph()
G.add_nodes_from(["A", "B", "C", "D", "E", "F", "G", "J", "K", "X", "Y", "Z"])
G.add_edges_from([
    ("A", "B"),
    ("B", "C"),
    ("C", "D"),
    ("D", "E"),
    ("E", "F"),
    ("F", "G"),
    ("G", "F"),
    ("G", "D"),
    ("J", "K"),
    ("X", "Y"),
    ("Y", "Z"),
    ("Z", "Y"),
])
```
I would like to find the following cycles:

1) D -> E -> F -> G -> D
2) F -> G -> F
3) Y -> Z -> Y

I believe the [`networkx.algorithms.cycles.find_cycle`](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cycles.find_cycle.html#networkx-algorithms-cycles-find-cycle) function is what I need, however, it only "Returns _a_ cycle found via depth-first traversal":

```python
nx.find_cycle(G)
>>> [('F', 'G'), ('G', 'F')]
nx.find_cycle(G, source="X")
>>> [('Y', 'Z'), ('Z', 'Y')]
```

So I am almost there, but I have two problems:

- How do I make sure I exhaustively search for every possible loop? The find misses the bigger loop (1) in the list above.
- In my real-scenario graph, I do not know _a priori_ which nodes are in which disconnected set of the graph, so it will be more difficult (although I guess not impossible) to manually select the `source` parameter of the `find_cycle` function. I guess this ties back to my previous question - if I can make sure that I exhaustively search throughout all the graph, then in theory I won't have to manually select the `source` parameter.

Any comments or ideas will be greatly appreciated.

**Comment by dschult** [comment, marked as answer]
I think you want the function `[cycle_basis](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cycles.cycle_basis.html)`.  This gives a set of cycles that can generate all possible cycles via addition/subtraction of the basis cycles (actually linear combination of the cycles) where addition means C1 + C2 traverses the nodes of C1 followed by the nodes of C2. 
Be careful -- there can be a *lot* of cycles.

Also `[simple_cycles](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cycles.simple_cycles.html)` might work better for you depending on what you are looking for (only the simple cycles).

**Reply by jeiros** [reply]
Thank you very much! The `simple_cycles` function meets the requirement 😄
